java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/UIDefaultsIconMetadata.java

```diff
@@ -0,0 +1,357 @@
+            LOG.log(Level.FINE, "Error wrapping default {0}: ", ok.toString());
+            LOG.log(Level.FINE, "Exception thrown", ex);
+        }
+        return o;
+    }
+    
+    private static void replace(UIDefaults defs, Object key, String type, Object replacement) {
+        LOG.log(Level.INFO, "Replaced {0}", key);
+        defs.put(key, replacement);
+    }
+    
+    public static final class MetadataImageIcon extends ImageIcon {
+        private final Icon delegateIcon;
+
+        public MetadataImageIcon(Icon delegateIcon, Image original) {
+            super(original);
+            this.delegateIcon = delegateIcon;
+        }
+
+        /**
+         * Get an {@link Icon} instance representing a scalable version of this {@code Image}.
+         *
+         * @return may be null
+         */
+        public Icon getDelegateIcon() {
+            return delegateIcon;
+        }
+
+        @Override
+        public int getIconHeight() {
+            return delegateIcon.getIconHeight();
+        }
+
+        @Override
+        public int getIconWidth() {
+            return delegateIcon.getIconWidth();
+        }
+
+        @Override
+        public void paintIcon(Component c, Graphics g, int x, int y) {
+            delegateIcon.paintIcon(c, g, x, y);
+        }
+    }
+    
+    /**
+     * Delegating Image which adds metadata to the original image's {@link Image#getProperty}.
+     * @author sdedic
+     */
+    public static class MetadataImage extends BufferedImage implements W {
+        private final Image original;
+
+        public MetadataImage(Image original, ColorModel cm,
+                          WritableRaster raster,
+                          boolean isRasterPremultiplied,
+                          Hashtable<?,?> properties) {
+            super(cm, raster, isRasterPremultiplied, properties);
+            this.original = original;
+        }
+
+        @Override
+        public Object getProperty(String name, ImageObserver observer) {
+            Object o = super.getProperty(name, observer);
+            return o != UndefinedProperty ? o : original.getProperty(name, observer);
+        }
+    }
+    
+    public static Object createNew(Image image, Icon icon, Hashtable props) {
+        ensureLoaded(image);
+        boolean bitmask = (image instanceof Transparency) && ((Transparency) image).getTransparency() != Transparency.TRANSLUCENT;
+        ColorModel model = colorModel(bitmask ? Transparency.BITMASK : Transparency.TRANSLUCENT);
+        int w = Math.max(1, image.getWidth(null));
+        int h = Math.max(1, image.getHeight(null));
+        if (icon == null && image instanceof Icon) {
+            icon = (Icon)image;
+        }
+        WritableRaster raster = model.createCompatibleWritableRaster(w, h);
+        BufferedImage newImage = new MetadataImage(image, model, raster, bitmask, props);
+
+        java.awt.Graphics g = newImage.createGraphics();
+        g.drawImage(image, 0, 0, null);
```

Comment:
This doesn't work for HiDPI icons or MultiResolutionImage? What kind of icons is this used for?

Reply:
In _this first_ version (which affects just NBLS headless distribution) - icons are exported from Nodes. Could you please suggest how to improve this part in the future so the HiDPI works OK ?

Reply:
> In this first version (which affects just NBLS headless distribution)
Is a future version of UIDefaultsIconMetadata supposed to actually run as part of the NetBeans IDE? Or is it always just in this headless mode?

(In any case, I'll have a look at the HiDPI implications here...)

Reply:
At this moment, we need to **replace** certain icons bcs their style does not match the LSP client's appearance. This does not apply to NB -- the icons would be simply changed or rebranded.

The client (vscode) has some icons already that use font glyphs, rebranding is not that much possible. Extracting glyphs, converting (e.g.) to SVGs for branding could bring some licensing issues I'd like to avoid. You're right (see other question) that the implementation is not optimal - this will hopefully improve.

Re. usage in NB apps that actually display the icons - I can think of possible usage for A11Y purposes, but did not went deep that way.